### ISSUE TITLE
Configure Cache Control Expires headers

### ISSUE THREAD

**Issue by nyurik** [maplibre/martin#749]
Martin should set some sort of the expiry HTTP headers, max-age or similar headers, next to the already existing `E-Tag` headers.

See also [MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control)

**Comment by CommanderStorm** [comment]
We likely can also make our caches expirable in a similar way -> https://docs.rs/moka/latest/moka/future/struct.CacheBuilder.html#method.expire_after

**Comment by nyurik** [comment]
i don't think we should mix internal cache and the expiration headers -- while they both speed up access, they function very differently, and have significantly different controls.  I would recommend https://www.varnish-software.com/developers/tutorials/object-lifetime/ as a good intro to how a proxy could behave - note that martin is somewhat of a proxy - esp if it gets content from a http-hosted pmtiles.

**Comment by CommanderStorm** [comment]
I guess you might have usecases where your internal cache is higher than the one you give out.
But still, this concept is very related and doing a full varnish configuration is likely overkill for a larger portion of users.

Not everyone is wikipedia and working with VCL is a resonably high learning curve 😉

**Comment by nyurik** [comment]
that's not what i meant.  Martin has many concepts similar to Varnish - so we may need at least some of the same cache controls to configure it:

* internal object cache
  *  TTL, Grace, Keep controls -- determine how long an object stays in Martin's internal cache, and what happens if the TTL has expired, .... This could be as simple as TTL only, but note that for max performance, we want to allow "respond while refreshing" - i.e. if a user asks for a tile, they may get an expired tile while Martin starts refreshing the tile cache in the background.  This way you don't get big delays in responses every time a popular tile is expired.
  * This should be discussed in a separate ticket
* Header values sent to the client - this is much simpler - we simply tell the client when to refresh.  This is what this ticket is about.

**Comment by CommanderStorm** [comment]
I am not sure if everything that you would in an ideal world want of the cache is currently possible with moka, but yea.. different issue.

**Comment by nyurik** [comment]
@CommanderStorm this ticket is not about internal Martin caching, only about what headers it sends to the clients.  Lets split the internal cache configuration into a separate ticket

**Comment by CommanderStorm** [comment]
I opened
- https://github.com/maplibre/martin/issues/2505 to discuss this, if someone wants to look into this 😉

**Comment by CommanderStorm** [comment]
From a chat on slack:

> We likely should have global header config, and per-source header configs
> i.e. we need a way to define:
> - root level cache headers for all tile-based and non-tile-based sources (fonts/sprites)
> - overrides per source type
> - override per group (e.g. all tables for a given connection, all pg functions, all files in a given directory)
> - override per individual source

What is currently unclear is the how this should be configured exactly

**Comment by AlexGacon** [comment]
Hi,

What is the status of the issue? Is it possible at least to have some basic cache header in the Martin response? Currently all my requests have a No-Cache value in the header. Is there a documentation on how to change that?

Regards
Alexandre

**Comment by CommanderStorm** [comment]
The status is that nobody has implemented it.
It is trivial and likely fun to implement if you want to have a go at it 😉

Currently, you will need to configure this via your reverse proxy, but we likely should also allow you to configure the header ^^

**Comment by AlexGacon** [comment]
It would be a pleasure if I had some basic knowledge in Rust ;).

**Comment by johncarmack1984** [comment]
I'd like to finish this one as well, again if that's alright. #3042 covered the global header; I have a [branch ready with the per-source half](https://github.com/maplibre/martin/compare/main...johncarmack1984:martin:john/cache-control-headers), following the convention #2673 established for per-source `cache`: a global default plus a `cache_control` key on each source config (file sources, postgres tables/functions, passthrough), validated at startup the same way as the global one.

Design notes:

- The value rides the existing per-source process-config channel and is applied in the tile handler, so martin-core stays out of HTTP concerns. Composite requests use it only when every requested source agrees; otherwise the server default applies.
- Of the four levels sketched above, this ships global + per-source. The resolver already supports a per-source-type tier, so that can follow separately if wanted.
- The global setting's test fixtures didn't survive the e2e migration in #3105, so I'd open this as two PRs: first restore e2e coverage of #3042, then the per-source feature.

I'm holding both until #3144 lands, since the reload path needs the same per-source carry-through that PR builds for `convert_to_*`. If that scope looks right I'll open them once it merges.
